test/integration/buffer_accounting_integration_test.cc

```diff
@@ -704,4 +716,263 @@ TEST_P(Http2OverloadManagerIntegrationTest, CanResetStreamIfEnvoyLevelStreamEnde
+  test_server_->waitForCounterEq("cluster.cluster_0.upstream_flow_control_resumed_reading_total",
+                                 1);
+}
+
+TEST_P(Http2DeferredProcessingIntegrationTest,
+       ShouldCloseDeferredUpstreamOnStreamCloseIfLocalReply) {
+  config_helper_.setBufferLimits(9000, 9000);
+  initialize();
+  Runtime::LoaderSingleton::getExisting()->mergeValues(
+      {{"envoy.reloadable_features.defer_processing_backedup_streams", "true"}});
+
+  // Stop writes to the downstream.
+  write_matcher_->setSourcePort(lookupPort("http"));
+  codec_client_ = makeHttpConnection(lookupPort("http"));
+  write_matcher_->setWriteReturnsEgain();
+
+  auto response_decoder = codec_client_->makeRequestWithBody(default_request_headers_, 1);
+  waitForNextUpstreamRequest();
+  FakeStreamPtr upstream_request = std::move(upstream_request_);
+  upstream_request->encodeHeaders(Http::TestResponseHeaderMapImpl{{":status", "200"}}, false);
+  upstream_request->encodeData(9000, false);
+
+  // Wait for an upstream response to have our reach its buffer limit and read
+  // disable.
+  test_server_->waitForGaugeEq("cluster.cluster_0.upstream_rq_active", 1);
+  test_server_->waitForCounterEq("cluster.cluster_0.upstream_flow_control_paused_reading_total", 1);
+
+  auto close_if_over_9000 =
+      [](StreamTee& tee, Http::StreamEncoderFilterCallbacks* encoder_cbs)
+          ABSL_EXCLUSIVE_LOCKS_REQUIRED(tee.mutex_) -> Http::FilterDataStatus {
+    if (tee.response_body_.length() > 9000) {
+      encoder_cbs->sendLocalReply(Http::Code::InternalServerError, "Response size was over 9000!",
+                                  nullptr, absl::nullopt, "");
+      return Http::FilterDataStatus::StopIterationNoBuffer;
+    }
+    return Http::FilterDataStatus::Continue;
+  };
+
+  EXPECT_TRUE(tee_filter_factory_.setEncodeDataCallback(0, close_if_over_9000));
+
+  upstream_request->encodeData(1, true);
+
+  // Verify codec received and has buffered onStreamClose for upstream as we're still read disabled.
+  buffer_factory_->waitUntilTotalBufferedExceeds(9001);
+  test_server_->waitForGaugeEq("cluster.cluster_0.http2.deferred_stream_close", 1);
+  test_server_->waitForCounterEq("cluster.cluster_0.upstream_flow_control_resumed_reading_total",
+                                 0);
+  EXPECT_TRUE(tee_filter_factory_.inspectStreamTee(0, [](const StreamTee& tee) {
+    absl::MutexLock l{&tee.mutex_};
+    EXPECT_EQ(tee.response_body_.length(), 9000);
+  }));
+
+  // Allow draining to the downstream, which should trigger a local reply.
+  write_matcher_->setResumeWrites();
+
+  ASSERT_TRUE(response_decoder->waitForReset());
+  ASSERT_TRUE(upstream_request->complete());
+  test_server_->waitForGaugeEq("cluster.cluster_0.http2.deferred_stream_close", 0);
+  test_server_->waitForCounterEq("cluster.cluster_0.upstream_flow_control_resumed_reading_total",
+                                 1);
+}
+
+TEST_P(Http2DeferredProcessingIntegrationTest,
```

Comment:
So I'd like to see 2 tests in addition to these integration test, and the two test are 
while data is backed up in the h2 buffers, some intermediate filter (unzip? but can be totally synthetic) adds enough data to have us go over the watermark limits and send a local response with [error response headers, reset]

I am 100% fine with a TODO for tese tests given the flag is off by default, but filter chain local response has been a common error mode and I want to make sure we handle this before we flip true.